docs/api/source/guibot.config.rst
guibot.config module
====================

.. automodule:: guibot.config
   :members:
   :undoc-members:
   :show-inheritance:


doc/tutorials/Match Settings Tutorial.md
**MATCH CONFIGURATION HIERARCHY**

In a simple case of image matching you can only define the path to the
needle and it will be performed for you. Example code for matching an
image */path/to/image\_folder/existing-image* is:

```
guibot.add_path('/path/to/image_folder')
guibot.find('existing-image')
```

However, one main disadvantage of this is that this matching requires
high default similarity and can be really fragile even if the slightest
change in the GUI occurs. You can of course use the calibrator to find
the match configuration for highest possible similarity in the new case
but it would be much better if you understand a little bit more about
how the match settings work and how modifying them a tiny bit can help
you achieve large flexibility of the *GUI scripts* and robustness to
insignificant changes in the GUI. So let's get started.

Next to any data file *existing-image.format* (here we assume image for
simplicity and therefore simpler backends) you can have a file to
contain its match settings with the name *existing-image.match*. If this
file exists, the match settings from it are automatically extracted and
used during the matching process of the respective image. Of course
there might be some cases where you need more specific match settings
so the match settings can always be processed further in the GUI
scripts you develop. Overall, this file ensures that you have reliable
long term configuration for matching the specific image. The file
however is specific for just one image. For all files without such
specific match settings, the defaults from the `GlobalConfig` object
are used. So all in all, the hierarchy of match configuration scopes
and overriding is the following:

1. `GlobalConfig` class - default settings for all targets every time
2. *target.match* file - default settings for a specific target every time
3. `target.match_settings` attribute - settings for a specific image
object (loaded image)
4. `region.finder` attribute - settings for a specific region

This allows for a kind of inheritance of the match configuration from
the most general to the most specific match cases. It is really necessary
since in real applications every time a needle image is matched, this
could be done with respect to different haystack images where there
might be different "distractions" relative to the actual match.

So to provide more details about the way to create more durable match
settings or temporary such we need to know the relevant API of the
classes `GlobalConfig`, `Target`, `Region`, and `Finder`. Every time you
perform matching you have to create `Image`, `Text`, `Pattern` or other
type of `Target` object depending on your preferred backend of use
(provided you have added the path to the image directory containing it)
and use that object as a needle image during matching. Of course you can
use that `Image` object multiple times in which case its last and most
current match configuration will be reused. If no match file was provided
the match settings will be the ones used for any matching and extracted
from the `GlobalConfig` class. Let's say that you need better match
settings for that specific image. For example if you want to change the
similarity required during each matching of this loaded image to 0.5 you
should do the following:

```
# run the next line once to use template default finder
GlobalConfig.find_backend = "template"
image = Image('existing-image')
image.use_own_settings = True
image.match_settings.params["find"]["similarity"].value = 0.5
```

We will explain more about the structure of the match settings a bit
later, this is just an example in order to see how to use the
`match_settings` attribute of the loaded image object. The new match
settings will be used in every matching where this `Image` object is
provided. You can be even more specific and change the match settings
for a single matching of the `Image` object (using `Region` as a more
ephemeral object than `Image`) by doing:

```
region = Region()
region.cv_backend.params["find"]["similarity"].value = 0.5
match = region.find('existing-image')
```

Notice however that in the long term this becomes cumbersome since you
would have to adapt the match configuration each time you load the image
or define a region. The way to solve this is to give more generality and
durability to the performed changes by saving them as a match file. You
can do so with:

```
image = Image('existing-image')
image.match_settings.params["find"]["similarity"].value = 0.5
image.use_own_settings = True
returned_image = image.save('/path/to/image_folder/existing-image.format')
```

This will additionally save a match file with the match configuration of
the image which will then be detected every next time you load the image:

```
image = Image('existing-image')
print image.use_own_settings
# True
print image.match_settings.params["find"]["similarity"].value
# 0.5
```

Are we good? All you need now is to know a bit more about the structure
of the match configuration and you are good to go.


**MATCH CONFIGURATION STRUCTURE**

Both the match configuration and backend implementation are carried out
within a *Finder* object. Knowing that, we can play with it more and go
beyond the hierarchy explained above by using custom match configuration
for multiple images and more. To create a few *Image* objects with the
same match settings you can do:

```
finder = TemplateFinder()
finder.params["find"]["similarity"].value = 0.5
image = Image('existing-image', match_settings=finder)
image2 = Image('existing-image2', match_settings=finder)
```

The configuration parameters or settings are organized into categories
(e.g. "find") and are objects of type `CVParameter`. The selection of
categories and parameters within each category depend on the choice of
computer vision backend. You can check the categories for a given backend
(i.e. `Finder` subclass) and the subalgorithm types they represent using

```
print finder.categories.keys()
# ['type', 'find', 'template']
print finder.categories
# {'type': 'backend_types', 'find': 'find_methods', 'template': 'template_matchers'}
```

You can then check the available and currently selected subalgorithms by

```
for category in finder.categories.keys():
    print finder.algorithms[finder.categories[category]]
    print "->", finder.params[category]["backend"]
# ('cv', 'dc')
# -> cv
# ('autopy', 'contour', 'template', 'feature', 'cascade', 'text', ...)
# -> template
# ('sqdiff_normed', 'ccorr_normed', 'ccoeff_normed')
# -> ccoeff_normed
```

To define a set of backend algorithms that you would like to use in
matching cases do:

```
finder.configure_backend(backend="sqdiff_normed", category="template")
```

or simply

```
finder.configure_backend(backend="sqdiff_normed")
```

Notice that if you replace a backend algorithm all relevant parameters
will be also replaced and if you set the same backend algorithm again
all relevant parameters will be set to the default. You can use a `reset`
flag to also rest all base category parameters but this is better
described in the API documentation of each backend. Parameters in a
base type are also parameters in a base category and are therefore more
general. In our example, the *similarity* parameter is more general, not
template spacific, and can be accessed from the base category *find* as

```
print finder.params["find"]["similarity"].value
# 0.5
```

An example of a less general parameter is *ScaleFactor* of a feature
finder. To access the *ScaleFactor* parameter of the ORB feature detector:

```
finder = FeatureFinder()
# optional if ORB is not the default feature detector
# finder.configure_backend(backend="ORB", category="fdetect")
# finder.synchronize_backend(backend="ORB", category="fdetect")
print finder.params["fdetect"]["ScaleFactor"].value
# 1.20000004768
```

Lastly, each parameter is a bit more complex than just a value. The
relevant attributes of a parameter are visible in any match file (which
of course is completely readable config file):

```
print finder.params["fdetect"]["ScaleFactor"]
<value='1.20000004768' min='1.01' max='2.0' delta='1.0' tolerance='0.1' fixed='True'>
```

From *min* and *max* we construct *range* tuple for the parameter object
which looks like

```
finder.params["fdetect"]["ScaleFactor"].range = (min, max)
```

If you don't care about calibration you don't need to know anything
about the other parameters. The *delta*, *tolerance*, and *fixed*
attributes are used by the calibrator where fixed would make the
calibrator skip calibrating the parameter. The other two would define
how much guessing will be used and when will the parameter have an
acceptable value for the calibrator.

The complexity behind the match configuration is a trade-off for the
availability of a unified configuration for all computer vision backends.
You need to read more about the specific computer vision algorithm to
know more about the specific backend parameters - this tutorial is
just to show you how to modify them and use the best methods you have
found for one or all of your images.


guibender/settings.py
# Copyright 2013 Intranet AG / Plamen Dimitrov and Thomas Jarosch
#
# guibender is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# guibender is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with guibender.  If not, see <http://www.gnu.org/licenses/>.
#

import re
import os
import logging
log = logging.getLogger('guibender.settings')

from errors import *


class GlobalSettings(type):
    """
    Metaclass used for the definition of static properties (the settings).

    We overwrite the name of the class in order to avoid documenting
    all settings here and adding an empty actual class. Instead, the resulting
    documentation contains just the settings class (using this as metaclass)
    and all settings respectively. In this way the front user should not worry
    about such implementation detail and simply use the provided properties.

    For those that like to think about it nonetheless: All methods of the
    resulting settings class are therefore static since they are methods of
    a class object, i.e. a metaclass instance.
    """

    # operational parameters shared between all instances
    _click_delay = 0.1
    _drag_delay = 0.5
    _drop_delay = 0.5
    _keys_delay = 0.2
    _type_delay = 0.1
    _rescan_speed_on_find = 0.2
    _screen_autoconnect = True
    _preprocess_special_chars = True
    _save_needle_on_error = True
    _image_logging_level = logging.ERROR
    _image_logging_destination = "./imglog"
    _image_logging_step_width = 3
    _image_quality = 3

    # backends shared between all instances
    _desktop_control_backend = "autopy"
    _find_backend = "hybrid"
    _contour_threshold_backend = "adaptive"
    _template_match_backend = "ccoeff_normed"
    _feature_detect_backend = "ORB"
    _feature_extract_backend = "ORB"
    _feature_match_backend = "BruteForce-Hamming"
    _text_detect_backend = "erstat"
    _text_ocr_backend = "tesseract"
    _hybrid_match_backend = "autopy"

    def click_delay(self, value=None):
        """
        Getter/setter for property attribute.

        :param value: time interval between two clicks in a double click
        :type value: float or None
        :returns: current value if no argument was passed otherwise only sets it
        :rtype: float or None
        """
        if value is None:
            return GlobalSettings._click_delay
        else:
            GlobalSettings._click_delay = value
    #: time interval between two clicks in a double click
    click_delay = property(fget=click_delay, fset=click_delay)

    def delay_after_drag(self, value=None):
        """
        Same as :py:func:`GlobalSettings.click_delay` but with

        :param value: timeout before drag operation
        """
        if value is None:
            return GlobalSettings._drag_delay
        else:
            GlobalSettings._drag_delay = value
    #: timeout before drag operation
    delay_after_drag = property(fget=delay_after_drag, fset=delay_after_drag)

    def delay_before_drop(self, value=None):
        """
        Same as :py:func:`GlobalSettings.click_delay` but with

        :param value: timeout before drop operation
        """
        if value is None:
            return GlobalSettings._drop_delay
        else:
            GlobalSettings._drop_delay = value
    #: timeout before drop operation
    delay_before_drop = property(fget=delay_before_drop, fset=delay_before_drop)

    def delay_before_keys(self, value=None):
        """
        Same as :py:func:`GlobalSettings.click_delay` but with

        :param value: timeout before key press operation
        """
        if value is None:
            return GlobalSettings._keys_delay
        else:
            GlobalSettings._keys_delay = value
    #: timeout before key press operation
    delay_before_keys = property(fget=delay_before_keys, fset=delay_before_keys)

    def delay_between_keys(self, value=None):
        """
        Same as :py:func:`GlobalSettings.click_delay` but with

        :param value: time interval between two consecutively typed keys
        """
        if value is None:
            return GlobalSettings._type_delay
        else:
            GlobalSettings._type_delay = value
    #: time interval between two consecutively typed keys
    delay_between_keys = property(fget=delay_between_keys, fset=delay_between_keys)

    def rescan_speed_on_find(self, value=None):
        """
        Same as :py:func:`GlobalSettings.click_delay` but with

        :param value: time interval between two image matching attempts
                      (used to reduce overhead on the CPU)
        """
        if value is None:
            return GlobalSettings._rescan_speed_on_find
        else:
            GlobalSettings._rescan_speed_on_find = value
    #: time interval between two image matching attempts (used to reduce overhead on the CPU)
    rescan_speed_on_find = property(fget=rescan_speed_on_find, fset=rescan_speed_on_find)

    def screen_autoconnect(self, value=None):
        """
        Getter/setter for property attribute.

        :param value: whether to perform complete initialization of the
                      desktop control backend
        :type value: bool or None
        :returns: current value if no argument was passed otherwise only sets it
        :rtype: bool or None
        :raises: :py:class:`ValueError` is value is not boolean or None

        Complete initialization includes connecting to the backend (screen)
        selected in the `_desktop_control_backend`.

        If disabled, you have to connect before performing any GUI operations::

            region.dc_backend.connect_screen()

        The use of this is to allow you to perform some configuration first.
        """
        if value is None:
            return GlobalSettings._screen_autoconnect
        elif value == True or value == False:
            GlobalSettings._screen_autoconnect = value
        else:
            raise ValueError
    #: whether to perform complete initialization of the desktop control backend
    screen_autoconnect = property(fget=screen_autoconnect, fset=screen_autoconnect)

    def preprocess_special_chars(self, value=None):
        """
        Same as :py:func:`GlobalSettings.screen_autoconnect` but with

        :param value: whether to preprocess capital and special characters and
                      handle them internally

        .. warning:: The characters will be forcefully preprocessed for the
            autopy on linux (capital and special) and vncdotool (capital) backends.
        """
        if value is None:
            return GlobalSettings._preprocess_special_chars
        elif value == True or value == False:
            GlobalSettings._preprocess_special_chars = value
        else:
            raise ValueError
    #: whether to preprocess capital and special characters and handle them internally
    preprocess_special_chars = property(fget=preprocess_special_chars, fset=preprocess_special_chars)

    def save_needle_on_error(self, value=None):
        """
        Same as :py:func:`GlobalSettings.screen_autoconnect` but with

        :param value: whether to perform an extra needle dump on matching error
        """
        if value is None:
            return GlobalSettings._save_needle_on_error
        elif value == True or value == False:
            GlobalSettings._save_needle_on_error = value
        else:
            raise ValueError
    #: whether to perform an extra needle dump on matching error
    save_needle_on_error = property(fget=save_needle_on_error, fset=save_needle_on_error)

    def image_logging_level(self, value=None):
        """
        Getter/setter for property attribute.

        :param value: logging level similar to the python logging module
        :type value: int or None
        :returns: current value if no argument was passed otherwise only sets it
        :rtype: int or None

        .. seealso:: See the image logging documentation for more details.
        """
        if value is None:
            return GlobalSettings._image_logging_level
        else:
            GlobalSettings._image_logging_level = value
    #: logging level similar to the python logging module
    image_logging_level = property(fget=image_logging_level, fset=image_logging_level)

    def image_logging_step_width(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_level` but with

        :param value: number of digits when enumerating the image
                      logging steps, e.g. value=3 for 001, 002, etc.
        """
        if value is None:
            return GlobalSettings._image_logging_step_width
        else:
            GlobalSettings._image_logging_step_width = value
    #: number of digits when enumerating the image logging steps, e.g. value=3 for 001, 002, etc.
    image_logging_step_width = property(fget=image_logging_step_width, fset=image_logging_step_width)

    def image_quality(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_level` but with

        :param value: quality of the image dumps ranging from 0 for no compression
                      to 9 for maximum compression (used to save space and reduce
                      the disk space needed for image logging)
        """
        if value is None:
            return GlobalSettings._image_quality
        else:
            GlobalSettings._image_quality = value
    #: quality of the image dumps ranging from 0 for no compression to 9 for maximum compression
    # (used to save space and reduce the disk space needed for image logging)
    image_quality = property(fget=image_quality, fset=image_quality)

    def image_logging_destination(self, value=None):
        """
        Getter/setter for property attribute.

        :param value: relative path of the image logging steps
        :type value: str or None
        :returns: current value if no argument was passed otherwise only sets it
        :rtype: str or None
        """
        if value is None:
            return GlobalSettings._image_logging_destination
        else:
            GlobalSettings._image_logging_destination = value
    #: relative path of the image logging steps
    image_logging_destination = property(fget=image_logging_destination, fset=image_logging_destination)

    def desktop_control_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the desktop control backend
        :raises: :py:class:`ValueError` if value is not among the supported backends

        Supported backends:
           * autopy - Windows, Linux (and OS X) compatible with both the GUI
                      actions and their calls executed on the same machine.
           * qemu - guest OS independent with GUI actions on a virtual machine
                    through Qemu Monitor object (provided by Autotest) and
                    their calls on the host machine.
           * vncdotool - guest OS independent or Linux remote OS with GUI
                         actions on a remote machine through vnc and their
                         calls on a vnc client machine.

        .. warning:: To use a particular backend you need to satisfy its dependencies,
            i.e. the backend has to be installed or you will have unsatisfied imports.
        """
        if value is None:
            return GlobalSettings._desktop_control_backend
        else:
            if value not in ["autopy", "qemu", "vncdotool"]:
                raise ValueError("Unsupported backend for GUI actions '%s'" % value)
            GlobalSettings._desktop_control_backend = value
    #: name of the desktop control backend
    desktop_control_backend = property(fget=desktop_control_backend, fset=desktop_control_backend)

    # these methods do not check for valid values since this
    # is already done during region and target initialization
    def find_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the computer vision backend

        Supported backends:
            * autopy - simple bitmap matching provided by autopy
            * contour - contour matching using overall shape estimation
            * template - template matching using correlation coefficients,
                         square difference, etc.
            * feature - matching using a mixture of feature detection,
                        extraction and matching algorithms
            * cascade - matching using OpenCV pretrained Haar cascades
            * text - text matching using ERStat or custom text detection,
                     followed by tesseract or Hidden Markov Model OCR
            * tempfeat - a mixture of template and feature matching where the
                       first is used as necessary and the second as sufficient stage
            * deep - deep learning matching using convolutional neural network but
                     customizable to any type of deep neural network
            * hybrid - use a composite approach with any of the above methods
                       as matching steps in a fallback sequence

        .. warning:: To use a particular backend you need to satisfy its dependencies,
            i.e. the backend has to be installed or you will have unsatisfied imports.
        """
        if value is None:
            return GlobalSettings._find_backend
        else:
            GlobalSettings._find_backend = value
    #: name of the computer vision backend
    find_backend = property(fget=find_backend, fset=find_backend)

    def contour_threshold_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the contour threshold backend

        Supported backends: normal, adaptive, canny.
        """
        if value is None:
            return GlobalSettings._contour_threshold_backend
        else:
            GlobalSettings._contour_threshold_backend = value
    #: name of the contour threshold backend
    contour_threshold_backend = property(fget=contour_threshold_backend, fset=contour_threshold_backend)

    def template_match_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the template matching backend

        Supported backends: autopy, sqdiff, ccorr, ccoeff, sqdiff_normed,
        ccorr_normed, ccoeff_normed.
        """
        if value is None:
            return GlobalSettings._template_match_backend
        else:
            GlobalSettings._template_match_backend = value
    #: name of the template matching backend
    template_match_backend = property(fget=template_match_backend, fset=template_match_backend)

    def feature_detect_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the feature detection backend

        Supported  backends: BruteForce, BruteForce-L1, BruteForce-Hamming,
        BruteForce-Hamming(2), in-house-raw, in-house-region.
        """
        if value is None:
            return GlobalSettings._feature_detect_backend
        else:
            GlobalSettings._feature_detect_backend = value
    #: name of the feature detection backend
    feature_detect_backend = property(fget=feature_detect_backend, fset=feature_detect_backend)

    def feature_extract_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the feature extraction backend

        Supported backends: ORB, FAST, STAR, GFTT, HARRIS, Dense, oldSURF.
        """
        if value is None:
            return GlobalSettings._feature_extract_backend
        else:
            GlobalSettings._feature_extract_backend = value
    #: name of the feature extraction backend
    feature_extract_backend = property(fget=feature_extract_backend, fset=feature_extract_backend)

    def feature_match_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the feature matching backend

        Supported backends: ORB, BRIEF, FREAK.
        """
        if value is None:
            return GlobalSettings._feature_match_backend
        else:
            GlobalSettings._feature_match_backend = value
    #: name of the feature matching backend
    feature_match_backend = property(fget=feature_match_backend, fset=feature_match_backend)

    def text_detect_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the text detection backend

        Supported backends: erstat, contours, components.
        """
        if value is None:
            return GlobalSettings._text_detect_backend
        else:
            GlobalSettings._text_detect_backend = value
    #: name of the text detection backend
    text_detect_backend = property(fget=text_detect_backend, fset=text_detect_backend)

    def text_ocr_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the optical character recognition backend

        Supported backends: tesseract, hmm, beamSearch.
        """
        if value is None:
            return GlobalSettings._text_ocr_backend
        else:
            GlobalSettings._text_ocr_backend = value
    #: name of the optical character recognition backend
    text_ocr_backend = property(fget=text_ocr_backend, fset=text_ocr_backend)

    def hybrid_match_backend(self, value=None):
        """
        Same as :py:func:`GlobalSettings.image_logging_destination` but with

        :param value: name of the hybrid matching backend for unconfigured one-step targets

        Supported backends: all nonhybrid backends of :py:func:`GlobalSettings.find_backend`.
        """
        if value is None:
            return GlobalSettings._hybrid_match_backend
        else:
            GlobalSettings._hybrid_match_backend = value
    #: name of the hybrid matching backend for unconfigured one-step targets
    hybrid_match_backend = property(fget=hybrid_match_backend, fset=hybrid_match_backend)


class GlobalSettings(object):
    """
    Handler for default configuration present in all
    cases where no specific value is set.

    The methods of this class are shared among
    all of its instances.
    """
    __metaclass__ = GlobalSettings


class LocalSettings(object):
    """
    Container for the configuration of all desktop control and
    computer vision backends, responsible for making them behave
    according to the selected parameters as well as for providing
    information about them and the current parameters.
    """

    def __init__(self, configure=True, synchronize=True):
        """
        Build a container for the entire backend configuration.

        :param bool configure: whether to also generate configuration
        :param bool synchronize: whether to also apply configuration

        Available algorithms can be seen in the `algorithms` attribute
        whose keys are the algorithm types and values are the members of
        these types. The algorithm types are shortened as `categories`.

        A parameter can be accessed as follows (example)::

            print self.params["control"]["vnc_hostname"]
        """
        self.categories = {}
        self.algorithms = {}
        self.params = {}

        self.categories["type"] = "backend_types"
        self.algorithms["backend_types"] = ("cv", "dc")

        if configure:
            self.__configure_backend()
        if synchronize:
            self.__synchronize_backend()

    def __configure_backend(self, backend=None, category="type", reset=False):
        if category != "type":
            raise UnsupportedBackendError("Backend category '%s' is not supported" % category)
        if reset:
            # reset makes no sense here since this is the base configuration
            pass
        if backend is None:
            backend = "cv"
        if backend not in self.algorithms[self.categories[category]]:
            raise UnsupportedBackendError("Backend '%s' is not among the supported ones: "
                                          "%s" % (backend, self.algorithms[self.categories[category]]))

        self.params[category] = {}
        self.params[category]["backend"] = backend

    def configure_backend(self, backend=None, category="type", reset=False):
        """
        Generate configuration dictionary for a given backend.

        :param backend: name of a preselected backend, see `algorithms[category]`
        :type backend: str or None
        :param str category: category for the backend, see `algorithms.keys()`
        :param bool reset: whether to (re)set all parent configurations as well
        :raises: :py:class:`UnsupportedBackendError` if `backend` is not among
                 the supported backends for the category (and is not `None`) or
                 the category is not found
        """
        self.__configure_backend(backend, category, reset)

    def configure(self, reset=True):
        """
        Generate configuration dictionary for all backends.

        :param bool reset: whether to (re)set all parent configurations as well

        If multiple categories are available and just some of them are configured,
        the rest will be reset to defaults. To configure specific category without
        changing others, use :py:function:`configure`.
        """
        self.configure_backend(reset=reset)

    def __synchronize_backend(self, backend=None, category="type", reset=False):
        if category != "type":
            raise UnsupportedBackendError("Backend category '%s' is not supported" % category)
        if reset:
            # reset makes no sense here since this is the base configuration
            pass
        # no backend object to sync to
        backend = "cv" if backend is None else backend
        if backend not in self.algorithms[self.categories[category]]:
            raise UninitializedBackendError("Backend '%s' has not been configured yet" % backend)

    def synchronize_backend(self, backend=None, category="type", reset=False):
        """
        Synchronize a category backend with the equalizer configuration.

        :param backend: name of a preselected backend, see `algorithms[category]`
        :type backend: str or None
        :param str category: category for the backend, see `algorithms.keys()`
        :param bool reset: whether to (re)sync all parent backends as well
        :raises: :py:class:`UnsupportedBackendError` if  the category is not found
        :raises: :py:class:`UninitializedBackendError` if there is no backend object
                 that is configured with and with the required name
        """
        self.__synchronize_backend(backend, category, reset)

    def synchronize(self, reset=True):
        """
        Synchronize all backends with the current configuration dictionary.

        :param bool reset: whether to (re)sync all parent backends as well
        """
        self.synchronize_backend(reset=reset)
